### PR TITLE
Convert DateFieldMapper to parametrized format

### PR DESCRIPTION
**Pull request by romseygeek**
This commit makes DateFieldMapper extend ParametrizedFieldMapper,
declaring its parameters explicitly.  As well as changes to DateFieldMapper
itself, there are some changes to dynamic mapping code to ensure that
dynamically detected date formats are passed through to new date mapper
builders.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java [hunk 1]
```diff
@@ -178,79 +159,64 @@ public static Resolution ofOrdinal(int ord) {
...
+            if (dateFormatter != null) {
+                this.format.setValue(dateFormatter.pattern());
+                this.locale.setValue(dateFormatter.locale());
+            }
         }
 
-        public boolean isFormatterSet() {
-            return format.explicit();
+        // for testing
+        public Builder(String name) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:198]: I'd prefer to drop these if we can manage it. Or to rename it to `public static Builder millisecondResolution(String name)` or something.

File: server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java [hunk 2]
```diff
@@ -311,6 +254,14 @@ public DateFieldType(String name) {
             this(name, true, true, DEFAULT_DATE_TIME_FORMATTER, Resolution.MILLISECONDS, Collections.emptyMap());
         }
 
+        public DateFieldType(String name, DateFormatter dateFormatter) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:257]: Do you think it'd be cleaner to use the builder to get these in tests?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 3]
```diff
@@ -681,17 +679,15 @@ private static void parseNullValue(ParseContext context, ObjectMapper parentMapp
...
                     }
-                    Mapper.Builder builder = context.root().findTemplateBuilder(context, currentFieldName, XContentFieldType.DATE);
+                    Mapper.Builder builder
+                        = context.root().findTemplateBuilder(context, currentFieldName, XContentFieldType.DATE, dateTimeFormatter);
                     if (builder == null) {
-                        builder = newDateBuilder(currentFieldName, dateTimeFormatter);
-                    }
-                    if (builder instanceof DateFieldMapper.Builder) {
-                        DateFieldMapper.Builder dateBuilder = (DateFieldMapper.Builder) builder;
-                        if (dateBuilder.isFormatterSet() == false) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:690]: :heart:

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 4]
```diff
@@ -122,6 +129,14 @@ public Version indexVersionCreated() {
                 return queryShardContextSupplier;
             }
 
+            public DateFormatter getDateFormatter() {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:132]: Probably worth javadoc - this is the format used for dynamic date detection and the default date format for date fields that don't have one set. Right?

**Review by nik9000 — APPROVED** [review]
I left some small things worth adding but it LGTM.

File: server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:253]: Builders give us field mappers though, not field types, so it feels like an extra unnecessary step.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:253]: Yeah. I can see that. I'm just not a big fan of having a bunch of ctors, especially when we have a fancy builder system already that makes it super clear what you are doing. Oh well. Its cool.

**Review by nik9000 — APPROVED** [review]
LGTM

**Pull request merged**